docs(websoc): add openapi specifications to schema

vinnyho opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description

- Added openapi query & response specifications in `websoc.ts` based off of `larc.ts`

<!--- Describe your changes in detail -->

## Related Issue

Closes #79 
## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->
- Improve API documentation by adding openapi specifications for WebSoc endpoints
- Makes it easier for API consumers and developers 
## How Has This Been Tested?

- Ran locally with changes

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] My code involves a change to the database schema.
- [ ] My code requires a change to the documentation.

laggycomputer commented on apps/api/src/schema/websoc.ts:
```diff
@@ -75,18 +75,37 @@ const isValidRestrictionCode = (code: string): code is (typeof restrictionCodes)
+    .openapi({ description: "The academic year for course search", example: "2025" }),
+  quarter: z.enum(terms).openapi({
+    description:
+      "The academic quarter 'Fall', 'Winter', 'Spring', 'Summer1', 'Summer10wk', or 'Summer2'",
```
Giving a description for enum terms which lists variants is redundant since Scalar already does so. In general, make sure you're checking the auto-generated `/reference` to see what your changes look like.

laggycomputer commented on apps/api/src/schema/websoc.ts:
```diff
@@ -118,30 +137,71 @@ export const websocQuerySchema = z.object({
     .optional()
+    .openapi({
+      description:
+        "The type of course level of the course (e.g. Lower Division, Upper Division, or Graduate)",
```
Same comment as earlier. For future reference, it would be misleading to give an example ("Upper Division") which does not match the actual string returned by the API ("UpperDiv"). Since the variants are already specified, don't worry about doing it again.

laggycomputer commented on apps/api/src/schema/websoc.ts:
```diff
@@ -118,30 +137,71 @@ export const websocQuerySchema = z.object({
     .union([z.enum(anyArray), z.enum(websocSectionTypes)])
     .optional()
+    .openapi({
+      description: "The type of section of the course (e.g. Lecture, Discussion, Lab)",
```
Same comment as course level field.

laggycomputer commented on apps/api/src/schema/websoc.ts:
```diff
@@ -118,30 +137,71 @@ export const websocQuerySchema = z.object({
     .enum(fullCoursesOptions)
     .optional()
+    .openapi({
+      description: "The enrollment status of the course",
```
This description is a bit unclear to me and since it's a query parameter, it shouldn't be phrased like it's a field of one course. Something like "How to filter courses by filled status" might be appropriate; workshop it and see what sounds right.

laggycomputer commented on apps/api/src/schema/websoc.ts:
```diff
@@ -118,30 +137,71 @@ export const websocQuerySchema = z.object({
-  startTime: timeSchema.optional(),
-  endTime: timeSchema.optional(),
+  cancelledCourses: z.enum(cancelledCoursesOptions).optional().openapi({
+    description: "Include or exclude cancelled courses",
```
You may want to consider rewording this to acknowledge the "cancelled class only" option though that's a more minor change.

laggycomputer commented on apps/api/src/schema/websoc.ts:
```diff
@@ -118,30 +137,71 @@ export const websocQuerySchema = z.object({
+    example: "Exclude",
+  }),
+  units: z.optional(z.literal("VAR").or(z.string())).openapi({
+    description: "The number of units for the courses",
```
Same comment as above; you might want to say something like "Filter courses by unit count". You should also explain how "VAR" is handled to fully explain this parameter. Take a look at the [source code](https://github.com/icssc/anteater-api/blob/2250280db9d667be91299d74d67fbdc0ad589aec/apps/api/src/services/websoc.ts#L201) to learn what it does.

laggycomputer commented on apps/api/src/schema/websoc.ts:
```diff
@@ -118,30 +137,71 @@ export const websocQuerySchema = z.object({
+    example: "4",
+  }),
+  startTime: timeSchema.optional().openapi({
+    description: "The course start time",
```
Same comments as previously; this is not the start time of a course but a parameter by which to filter an unspecified number of courses. Additionally, you should look at the [source code](https://github.com/icssc/anteater-api/blob/2250280db9d667be91299d74d67fbdc0ad589aec/apps/api/src/services/websoc.ts#L208) to see what this does, since this explanation of what the parameter does is incorrect.

laggycomputer commented on apps/api/src/schema/websoc.ts:
```diff
@@ -118,30 +137,71 @@ export const websocQuerySchema = z.object({
+    example: "08:00",
+  }),
+  endTime: timeSchema.optional().openapi({
+    description: "The course end time",
```
Same comment as the one concerning start time.

laggycomputer commented on apps/api/src/schema/websoc.ts:
```diff
@@ -118,30 +137,71 @@ export const websocQuerySchema = z.object({
     .string()
     .optional()
+    .openapi({
+      description: "Exclude courses with specific restriction codes",
```
You may want to state precisely that this list should be comma-separated, since that is the language developers understand. It would also be a nice addition to link to the official registrar's page explaining what these codes mean (though not strictly needed).

laggycomputer reviewed (changes requested):
Great progress! There are some inaccuracies that need addressing, but clean those up and these are great additions.

vinnyho commented:
Thanks for the suggestions, I apologize you had to leave a lot of suggestions on the PR that was bad of me

I added your suggested fixes, and added more clarification by including 'WebSoc' in the descriptions as `larc.ts` did. Please let me know if there are any other fixes necessary. Thanks!

laggycomputer commented on apps/api/src/schema/websoc.ts:
```diff
@@ -75,18 +75,36 @@ const isValidRestrictionCode = (code: string): code is (typeof restrictionCodes)
-    .length(4, { message: "Parameter 'year' must have length 4" }),
-  quarter: z.enum(terms),
+    .length(4, { message: "Parameter 'year' must have length 4" })
+    .openapi({ description: "The year of the WebSoc section", example: "2025" }),
```
Adding "WebSoc" every time is clunky and it should be clear from surrounding context that this query parameter is undeniably for WebSoc. You should also look closely at the results of this endpoint and determine where "course" or "section" is the correct term. I will not repeat that comment but it extends throughout. When #73 is resolved, this schema definition in particular will become obsolete.

laggycomputer commented on apps/api/src/schema/websoc.ts:
```diff
@@ -118,30 +136,73 @@ export const websocQuerySchema = z.object({
+  }),
+  days: daysSchema.optional().openapi({
+    description: "Selects WebSoc sections held on any of the given days",
+    example: "M,Tu,W,Th",
```
If convenient, consider demonstrating more day aliases as defined in the day schema. My previous comment about expressly articulating comma separation stands.

laggycomputer commented on apps/api/src/schema/websoc.ts:
```diff
@@ -118,30 +136,73 @@ export const websocQuerySchema = z.object({
+    example: "ALP",
+  }),
+  room: z.string().optional().openapi({
+    description: "The room of the WebSoc section",
```
"room number" is more precise.

laggycomputer commented on apps/api/src/schema/websoc.ts:
```diff
@@ -118,30 +136,73 @@ export const websocQuerySchema = z.object({
     .union([z.enum(anyArray), z.enum(websocSectionTypes)])
     .optional()
+    .openapi({
+      description: "The section type of the WebSoc section, see [Registrar](https://reg.uci.edu/)",
```
If you link a page at all, it should be the exact page which describes the meaning of these codes. Linking the registrar's homepage is not helpful.

laggycomputer commented on apps/api/src/schema/websoc.ts:
```diff
@@ -118,30 +136,73 @@ export const websocQuerySchema = z.object({
-  endTime: timeSchema.optional(),
+  cancelledCourses: z.enum(cancelledCoursesOptions).optional().openapi({
+    description:
+      "Filter courses based on their cancellation status, including only cancelled courses",
```
I see you've addressed the comment from earlier but the resulting description is vague. A better option may be "Exclude, include, or only include...".

laggycomputer commented on apps/api/src/schema/websoc.ts:
```diff
@@ -118,30 +136,73 @@ export const websocQuerySchema = z.object({
+  }),
+  units: z.optional(z.literal("VAR").or(z.string())).openapi({
+    description:
+      "Filter WebSoc sections by unit count. The value 'VAR' is used for courses with variable unit values, where the unit count can vary depending on the course section",
```
This description can be more concise, e.g. "Filter for a specific unit count". You should also look carefully at the query parsing code to explain as much if it as you can in this description.

laggycomputer commented on apps/api/src/schema/websoc.ts:
```diff
@@ -118,30 +136,73 @@ export const websocQuerySchema = z.object({
+    example: "4",
+  }),
+  startTime: timeSchema.optional().openapi({
+    description: "Filters WebSoc sections for courses that start at or after this time",
```
This can be more concise. In particular saying both "sections" and "courses" produces added confusion.

laggycomputer commented on apps/api/src/schema/websoc.ts:
```diff
@@ -118,30 +136,73 @@ export const websocQuerySchema = z.object({
+  }),
+  endTime: timeSchema.optional().openapi({
+    description: "Filters WebSoc sections for courses that end at or before this time",
+    example: "17:00",
```
If 24-hour time is indeed the format expected by the code (check this yourself), you should say so in both descriptions.

laggycomputer commented on apps/api/src/schema/websoc.ts:
```diff
@@ -118,30 +136,73 @@ export const websocQuerySchema = z.object({
     .optional()
+    .openapi({
+      description:
+        "Excludes WebSoc sections by comma-separated restriction codes, see [Restriction Codes](https://www.reg.uci.edu/enrollment/restrict_codes.html)",
```
The display name of this link looks out of place and should more clearly label where it is pointing, e.g. "Registrar's list" or "<site> for list of codes".

laggycomputer commented:
Please also ensure you resolve the merge conflict, mark previous comments you've fixed as unresolved, and rename the PR title pursuant to the [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) standard. Your PR will be squashed and rebased onto `main` using the PR title as the commit title and we require such commit titles follow conventional commits.

andrew-wang0 commented:
@vinnyho looks like a lot of good progress so far! One small nitpick, could you rename the PR title to follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary)

Something like `feat(websoc): add openapi specifications to schema`

vinnyho commented:
The merge conflict conflict seems to be happening as I add the description and example for quarter and year, I don't think I can resolve it without just removing what I added.

laggycomputer commented:
> The merge conflict conflict seems to be happening as I add the description and example for quarter and year, I don't think I can resolve it without just removing what I added.

A correct merge conflict resolution is an actual merge commit with the tip of your PR branch as first parent and `origin/main` as the second parent. You need to make this merge commit yourself.

vinnyho commented:
I believe I fixed the issue, but I don't know for sure. This is my first time dealing with merging conflicts as I didn't come across it with other PRs. Please let me know if I did anything wrong thank you

laggycomputer commented on apps/api/src/schema/lib/day.ts:
```diff
@@ -80,5 +80,6 @@ const daysTransform = (days: string | undefined, ctx: z.RefinementCtx) => {
 };
 
 export const daysSchema = z.string().transform(daysTransform).openapi({
+  description: "The days of the week for when the course meets",
```
This is not quite right. Try this endpoint yourself and see what this day parameter does. The way it's worded also seems to describe a course returned by the endpoint, whereas you are describing a filter \*parameter* on the query which could apply to any number of returned objects.

I also assume you've already verified that this description is correct for LARC queries as well, the other place where this schema is used.

laggycomputer commented on apps/api/src/schema/websoc.ts:
```diff
@@ -75,18 +75,36 @@ const isValidRestrictionCode = (code: string): code is (typeof restrictionCodes)
-    .length(4, { message: "Parameter 'year' must have length 4" }),
-  quarter: z.enum(terms, { required_error: "Parameter 'quarter' is required" }),
+    .length(4, { message: "Parameter 'year' must have length 4" })
+    .openapi({ description: "The academic year of the course ", example: "2025" }),
```
Trailing space.

laggycomputer commented on apps/api/src/schema/websoc.ts:
```diff
@@ -75,18 +75,36 @@ const isValidRestrictionCode = (code: string): code is (typeof restrictionCodes)
+    .length(4, { message: "Parameter 'year' must have length 4" })
+    .openapi({ description: "The academic year of the course ", example: "2025" }),
+  quarter: z.enum(terms, { required_error: "Parameter 'quarter' is required" }).openapi({
+    description: "The academic quarter of the course",
```
Again, this is worded like it's a property of a single course returned by the endpoint when it really exists in the query to filter an unknown number of courses. Think of this as one of the most broad parameters of the query. Something like: "The quarter in which to search".

laggycomputer commented on apps/api/src/schema/websoc.ts:
```diff
@@ -75,18 +75,36 @@ const isValidRestrictionCode = (code: string): code is (typeof restrictionCodes)
   ge: z
     .enum(geCategories)
     .optional()
+    .openapi({ description: "The general education category of the course", example: "GE-1A" })
```
Comment about singular course and notion of query search parameter vis-a-vis response subfield stands throughout this review.

laggycomputer commented on apps/api/src/schema/websoc.ts:
```diff
@@ -75,18 +75,36 @@ const isValidRestrictionCode = (code: string): code is (typeof restrictionCodes)
+    example: "DATA STRC IMPL&ANLS",
+  }),
+  courseNumber: courseNumberSchema.optional().openapi({
+    description: "The course number(s) of the course",
```
You should document how comma-separation and course ranges are possible, probably on the definition of `courseNumberSchema`.

laggycomputer commented on apps/api/src/schema/websoc.ts:
```diff
@@ -75,18 +75,36 @@ const isValidRestrictionCode = (code: string): code is (typeof restrictionCodes)
+  }),
+  courseNumber: courseNumberSchema.optional().openapi({
+    description: "The course number(s) of the course",
+    example: "46",
```
Overriding the original [example](https://github.com/icssc/anteater-api/blob/eb7773fde10dfee159531ef06d9b618021e855a7/apps/api/src/schema/lib/course.ts#L52) of `courseNumberSchema` in this way omits information about the breadth of what this parameter can do.

laggycomputer commented on apps/api/src/schema/websoc.ts:
```diff
@@ -75,18 +75,36 @@ const isValidRestrictionCode = (code: string): code is (typeof restrictionCodes)
     .string()
     .optional()
+    .openapi({
+      description: "The section code of the course",
```
Look at the code just below this patch and the choice of plural in `sectionCodes`. This parameter does more than getting exactly one section code. Describe that behavior sufficiently for a specification.

Also note that section codes describe sections, not codes.

laggycomputer commented on apps/api/src/schema/websoc.ts:
```diff
@@ -75,18 +75,36 @@ const isValidRestrictionCode = (code: string): code is (typeof restrictionCodes)
     .optional()
+    .openapi({
+      description: "The section code of the course",
+      example: "35530",
```
The example should be expanded accordingly.

laggycomputer commented on apps/api/src/schema/websoc.ts:
```diff
@@ -118,30 +136,69 @@ export const websocQuerySchema = z.object({
     .union([z.enum(anyArray), z.enum(websocSectionTypes)])
     .optional()
+    .openapi({
+      description: "Filter courses by section type",
```
Expanding on the comment extending throughout this review: A section type describes a section; a course as a whole is not a lecture or recitation. In fact, a course can have sections of only one type or of multiple. Then it doesn't make sense to claim you can "filter courses by section type". Reword accordingly.

laggycomputer commented on apps/api/src/schema/websoc.ts:
```diff
@@ -118,30 +136,69 @@ export const websocQuerySchema = z.object({
+  }),
+  units: z.optional(z.literal("VAR").or(z.string())).openapi({
+    description:
+      "Filter courses by unit count. 'VAR' filters for courses with variable unit ranges (e.g., '1-4')",
```
That is not all of the functionality of this parameter.

laggycomputer commented on apps/api/src/schema/websoc.ts:
```diff
@@ -118,30 +136,69 @@ export const websocQuerySchema = z.object({
     .optional()
+    .openapi({
+      description:
+        "Exclude courses by comma-separated restriction codes, see [Registrar's list](https://www.reg.uci.edu/enrollment/restrict_codes.html)",
```
Earlier comment about a different display name for this link stands. You also need to clearly indicate in text that this is a link; look at your local copy of the REST reference to see why the current description is problematic.

laggycomputer commented on openapi:
Not sure what this is but it should not be in the source tree. Delete it on your branch.

laggycomputer commented on apps/api/src/schema/websoc.ts:
```diff
@@ -118,30 +136,69 @@ export const websocQuerySchema = z.object({
+    description: "The instructor's name for the course",
+    example: "Anteater, P.",
+  }),
+  days: daysSchema.optional().openapi({}),
```
`.openapi` is not needed.

ecxyzzy commented:
Closing as this PR has had no activity for the past two weeks.
